src/ripple/overlay/impl/PeerImp.cpp

```diff
@@ -784,13 +785,19 @@ PeerImp::doAccept()
         return buf;
     }();
 
+    // Put a placeholder in the send queue so no other messages will be
+    // sent until this one is complete.
+    send_queue_.push({});
```

Comment:
This is going to block the queue permanently. If there is another `send()` before the handshake response callback is called then the message from this send is added to the send queue since the queue is not empty. The queue now has the dummy message and the valid message. When the queue is popped in the callback and the dummy message is removed, the valid message will permanently block the queue. A way to do this is to call `onWriteMessage({},0)` in the callback to kick-start the queue processing. This has to be done on the strand to synchronize access to the send queue. How about just adding a flag `protocolStarted_` initially set to `false`. The flag is set to `true` in `doProtocolStart()` before any message is sent. Then we can assert in the `send()` if the flag is false. This is a simpler solution and in general we should not send any message before the handshake is complete. Also the dummy message doesn't address a case, albeit hypothetical, when a message is sent before the dummy message is added to the queue.